plugins/gssapi.c

```diff
@@ -811,6 +829,229 @@ gssapi_server_mech_new(void *glob_context,
+      ccname = NULL;
+    }
+
+    switch (needle[1]) {
+    case 'u': {
+#ifdef HAVE_GETPWNAM
+      char uid_str[100];
+      struct passwd* pwd = getpwnam(username);
+      sprintf(uid_str, "%d", pwd->pw_uid);
+      if (NULL == (ccname = strsubst(text, ccache_copy, "%u", uid_str))) {
+        goto parsing_failed;
+      }
+#else
+      text->utils->seterror(text->utils->conn, SASL_LOG_ERR,
+                            "GSSAPI error: %u is supported in ccache");
+      goto parsing_failed;
+#endif
+      break;
+    }
+    case 'U': {
+      ccname = strsubst(text, ccache_copy, "%U", username);
+      break;
+    }
+    case 'e': {
+#ifdef HAVE_GETEUID
+      char euid_str[100];
+      uid_t euid = geteuid();
+      sprintf(euid_str, "%ul", euid);
+      if (NULL == (ccname = strsubst(text, ccache_copy, "%e", euid_str))) {
+        goto parsing_failed;
+      }
+#else
+      text->utils->seterror(text->utils->conn, SASL_LOG_ERR,
+                            "GSSAPI error: %%e is supported in ccache");
+      goto parsing_failed;
+#endif
+      break;
+    }
+    case 'E': {
+#ifdef HAVE_GETLOGIN_R
+      char server_username[100];
+      getlogin_r(server_username, sizeof(server_username));
+      if (NULL == (ccname = strsubst(text, ccache_copy, "%E", server_username))) {
+        goto parsing_failed;
+      }
+#else
+      text->utils->seterror(text->utils->conn, SASL_LOG_ERR,
+                            "GSSAPI error: %%E is supported in ccache");
+      goto parsing_failed;
+#endif
+      break;
+    }
+    case 'p': {
+#ifdef HAVE_GETPID
+      char pid_str[100];
+      pid_t pid;
+      pid = getpid();
+      sprintf(pid_str, "%d", pid);
+      if (NULL == (ccname = strsubst(text, ccache_copy, "%p", pid_str))) {
+        goto parsing_failed;
+      }
+#else
+      text->utils->seterror(text->utils->conn, SASL_LOG_ERR,
+                            "GSSAPI error: %%p is supported in ccache");
+      goto parsing_failed;
+#endif
+      break;
+    }
+    default:
+      text->utils->seterror(text->utils->conn, SASL_LOG_ERR,
+                            "GSSAPI error: invalid ccache format");
+      goto invalid_ccache_format;
+      break;
+    }
+
+    free(ccache_copy);
+    if (NULL != ccname) {
+      ccache_copy = strdup(ccname);
+      if (NULL == ccache_copy) {
+        text->utils->seterror(text->utils->conn, SASL_LOG_ERR,
+                              "GSSAPI error: out of memory");
+        goto malloc_failed;
+      }
+    }
+  }
+
+  if ((NULL == ccname) && (NULL != ccache)) {
+    if (NULL == (ccname = strdup(ccache))) {
+      text->utils->seterror(text->utils->conn, SASL_LOG_ERR,
+                            "GSSAPI error: out of memory");
+      goto malloc_failed;
+    }
+  }
+
+  goto ok;
+ invalid_ccache_format:
+ malloc_failed:
+ sanity_check_failed:
+ parsing_failed:
+  if (NULL != ccname) {
+    free(ccname);
+    ccname = NULL;
+  }
+ noccache:
+  if (NULL == ccname) {
+    ccname = strdup(getenv("KRB5CCNAME"));
```

Comment:
Nothing guarantees a KRB5CCNAME is always set, libkrb5 does not depend on it, as it can use the default ccache defined in krb5.conf that can use substitutions too.
You should not fail if KRB5CCNAMe is empty and  just let the default ccache be used I think.

Reply:
I guess; what should be better is if this resolve-function wont succeed, the calling code should make the decision to resolve using the krb5.conf settings/default in that case somehow. Do you have a nice little suggestion on how to get hold of it if it's not in the ENV? I was kind-of hoping automagic fallback to the host default config.

Reply:
If your plan is to use the defaults, then do not do anything here.
gssapi/krb5 will always use KRB5CCNAME and fallback to krb5.conf profile options on its own already.

Reply:
Great, thanks. :+1:

Reply:
@simo5  I'm trying to make a sensible fall-back flow according to https://web.mit.edu/kerberos/krb5-1.12/doc/basic/ccache_def.html#default-ccache-name but I wanted to take the chance to ask you if you know of any API in the GSSAPI-part of KRB5 that would give me this automagically, have not found anything relevant.

Reply:
... Also... I'm thinking. The gss_store_cred_into (which I use) is relying on the configurable stuff a bit more. Maybe if there is a gss_store_cred() call as fallback and reduced API capabilities it should handle the fall-back flow, right?